wp/lib/bap_wp/src/output.ml

```diff
@@ -60,6 +60,48 @@ let format_model (model : Model.model) (env1 : Env.t) (_env2 : Env.t) : string =
   Format.flush_str_formatter ()
 
 
+type mem_model = {default : Z3.Expr.expr ; model : (Z3.Expr.expr * Z3.Expr.expr) list}
+
+(** [extract_array] takes a z3 expression that is a seqeunce of store and converts it into
+    a mem_model, which consists of a key/value association list and a default value *)
+
+let extract_array (e : Z3.Expr.expr) : mem_model  = 
+      let rec extract_array' (partial_map : (Z3.Expr.expr * Z3.Expr.expr) list) (e : Z3.Expr.expr) : mem_model =
+          Printf.printf "%s\n" (Z3.Expr.to_string e);
+          let numargs = Z3.Expr.get_num_args e in
+          let args = Z3.Expr.get_args e in
+          let f_decl = Z3.Expr.get_func_decl e in
+          let f_name = Z3.FuncDecl.get_name f_decl |> Z3.Symbol.to_string in
+          if ( Int.(numargs = 3) && String.(f_name = "store"))
+          then begin 
+                let next_arr = List.nth_exn args 0 in
+                let key = List.nth_exn args 1 in
+                let value = List.nth_exn args 2 in
+                extract_array' (( key , value ) :: partial_map) next_arr
+                end
+          else if ( Int.(numargs = 1) && String.(f_name = "const")) then begin
+                let key = List.nth_exn args 0 in
+                { default = key; model = List.rev partial_map}
+                end
+          else begin
+                Printf.printf "Unpexpected case destructing Z3 array: %s" f_name;
```

Comment:
Might want to make the a warning statement:
`warning "Unexpected ... %s\n" f_name;`